serving/src/main/java/feast/serving/service/OnlineServingService.java

```diff
@@ -64,110 +62,211 @@ public GetFeastServingInfoResponse getFeastServingInfo(
+   * @param isStaleValues whether which the feature row contains stale values.
+   * @return valueMap mapping string feature reference to feature valuyesaq
+   */
+  private Map<String, Value> unpackValueMap(
+      FeatureRow featureRow, FeatureSetRequest featureSetRequest, boolean isStaleValues) {
+    Map<String, Value> valueMap = new HashMap<>();
+    // In order to return values containing the same feature references provided by the user,
+    // we reuse the feature references in the request as the keys in field builder map
+    Map<String, FeatureReference> nameRefMap = featureSetRequest.getFeatureRefsByName();
+    if (featureRow != null) {
+      // unpack feature row's feature values and populate value map
+      Map<String, Value> featureValueMap =
+          featureRow.getFieldsList().stream()
+              .filter(featureRowField -> nameRefMap.containsKey(featureRowField.getName()))
+              .collect(
+                  Collectors.toMap(
+                      featureRowField -> {
+                        FeatureReference featureRef = nameRefMap.get(featureRowField.getName());
+                        return RefUtil.generateFeatureStringRef(featureRef);
+                      },
+                      featureRowField -> {
+                        // drop stale feature values.
+                        if (isStaleValues) {
+                          return Value.newBuilder().build();
+                        } else {
+                          return featureRowField.getValue();
+                        }
+                      }));
+      valueMap.putAll(featureValueMap);
+    }
 
-  private void populateRequestCountMetrics(FeatureSetRequest featureSetRequest) {
-    String project = featureSetRequest.getSpec().getProject();
-    featureSetRequest
-        .getFeatureReferences()
-        .parallelStream()
-        .forEach(ref -> Metrics.requestCount.labels(project, ref.getName()).inc());
-  }
+    // create empty values for features specified in request but not present in feature row.
+    Set<String> missingFeatures =
+        nameRefMap.values().stream()
+            .map(ref -> RefUtil.generateFeatureStringRef(ref))
+            .collect(Collectors.toSet());
+    missingFeatures.removeAll(valueMap.keySet());
+    missingFeatures.forEach(refString -> valueMap.put(refString, Value.newBuilder().build()));
 
-  @Override
-  public GetBatchFeaturesResponse getBatchFeatures(GetBatchFeaturesRequest getFeaturesRequest) {
-    throw Status.UNIMPLEMENTED.withDescription("Method not implemented").asRuntimeException();
+    return valueMap;
   }
 
-  @Override
-  public GetJobResponse getJob(GetJobRequest getJobRequest) {
-    throw Status.UNIMPLEMENTED.withDescription("Method not implemented").asRuntimeException();
+  /**
+   * Generate Field level Status metadata for the given valueMap.
+   *
+   * @param isNotFound whether the given valueMap represents values that were not found in the
```

Comment:
Why does the method have awareness of what is calling it? Wouldn't it make more sense for the caller to provide the status?